tomviz/TomographyReconstruction.cxx

```diff
@@ -0,0 +1,169 @@
+#include <math.h>
+#include "TomographyReconstruction.h"
+#include "vtkImageData.h"
+#include "vtkFieldData.h"
+#include "vtkDataArray.h"
+#define PI 3.14159265359
+#include "vtkPointData.h"
+#include "vtkFloatArray.h"
+#include "vtkSmartPointer.h"
+
+#include <QDebug>
+
+namespace tomviz
+{
+TomographyReconstruction::TomographyReconstruction()
+{
+
+}
+
+TomographyReconstruction::~TomographyReconstruction()
+{
+
+}
+
+// conversion code
+template<typename T>
+vtkSmartPointer<vtkFloatArray> convertToFloatT(T *data, int len)
+{
+  vtkSmartPointer<vtkFloatArray> array = vtkSmartPointer<vtkFloatArray>::New();
+  array->SetNumberOfTuples(len);
+  float *f = static_cast<float*>(array->GetVoidPointer(0));
+  for (int i = 0; i < len; ++i)
+  {
+    f[i] = (float) data[i];
+  }
+  return array;
+}
+  
+vtkSmartPointer<vtkFloatArray> convertToFloat(vtkImageData* image)
+{
+  vtkDataArray *scalars = image->GetPointData()->GetScalars();
+  int len = scalars->GetNumberOfTuples();
+  vtkSmartPointer<vtkFloatArray> array;
+  switch(scalars->GetDataType())
+  {
+    vtkTemplateMacro(array = convertToFloatT(static_cast<VTK_TT*>(scalars->GetVoidPointer(0)), len););
+  }
+  return array;
+}
+
+  
+//3D Weighted Back Projection reconstruction
+void TomographyReconstruction::weightedBackProjection3(vtkImageData *tiltSeries,vtkImageData *recon)
+{
+  int extents[6];
+  tiltSeries->GetExtent(extents);
+  int xDim = extents[1] - extents[0] + 1; //number of slices
+  int yDim = extents[3] - extents[2] + 1; //number of rays
+  int zDim = extents[5] - extents[4] + 1; //number of tilts
+  
+  //Get tilt angles
+  vtkDataArray *tiltAnglesArray = tiltSeries->GetFieldData()->GetArray("tilt_angles");
+  int numOfTilts = tiltAnglesArray->GetNumberOfTuples();
+  double *tiltAngles = static_cast<double*>(tiltAnglesArray->GetVoidPointer(0));
+  
+  // Creating the output volume and getting a pointer to it
+  int outputSize[3] = { xDim, yDim, yDim};
+  recon->SetExtent(0, outputSize[0] - 1, 0, outputSize[1] - 1, 0, outputSize[2] - 1);
+  recon->AllocateScalars(VTK_FLOAT, 1); // 1 is for one component (i.e. not vector)
+  float *reconPtr = static_cast<float*>(recon->GetScalarPointer());
+  
+  //Reconstruction
+  float *sinogram = new float[yDim*zDim]; //Placeholder for 2D sinogram
+  float *recon2d = new float[yDim*yDim]; //Placeholder for 2D reconstruction (y-z plane)
+  for (int s = 0; s < xDim; ++s) //loop through slices (x-direction)
+  {
+    //Get sinogram
+    TomographyReconstruction::tiltSeriesToSinogram(tiltSeries, s, sinogram);
+    //2D back projection
+    TomographyReconstruction::unweightedBackProjection2(sinogram, tiltAngles, recon2d, zDim, yDim);
+    //Put recon into
+    for (int iy = 0; iy < outputSize[1]; ++iy) //loop through all pixels in reconstructed image (y-z plane)
+      for (int iz = 0; iz < outputSize[2]; ++iz)
+      {
+        reconPtr[iz*outputSize[0]*outputSize[1] + iy*outputSize[0] + s] = recon2d[iy*outputSize[0] + iz];
+      }
+  }

```

Comment:
@mathturtle Line 87 to 102 shows how I currently do the 3D reconstruction.
First, I allocate two pointers as placeholders for sinograms and 2D reconstructions (line 88 and 89).
Next, I loop through each slice (y-z plane) in the tilt series (line 90). For each slice, I extract the corresponding sinogram (line 93) and use it to obtain a 2D reconstruction (line 95). Lastly, I copy the 2D reconstruction to the 3D volume (reconPtr). 

I think we can implement the alignment UI in similar fashion:
Every time user specify the a slice No., 
1. Extract the corresponding sinogram (using "tiltSeriesToSinogram")
2. Reconstruct the new sinogram (using "unweightedBackProjection2")
3. Update recon window in the UI.